Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,191 @@
+
+            // We know we're about to overwrite the oldest message, so advance the reader to the second oldest message.
+            try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: false, cacheOverwritesOldMessages: true)
+
+            // Start writing from the beginning of the file.
+            try write(messageData: messageData)
+
+        } else {
+            // We're out of room.
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+    }
+
+    /// Fetches all messages from the cache.
+    public func cachedMessages() throws -> [T] {
+        try setUpFileHandlesIfNecessary()
+
+        var messages = [T]()
+        while let encodedMessage = try reader.nextEncodedMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages) {
+            messages.append(try decoder.decode(T.self, from: encodedMessage))
+        }
+
+        return messages
+    }
+
+    // MARK: Private
+
+    /// Seeks the reader and writer file handles to the correct place.
+    /// Only necessary the first time this method is called.
+    private func setUpFileHandlesIfNecessary() throws {
+        guard !hasSetUpFileHandles else {
+            return
+        }
+
+        // This is our first cache action.
+        // We need to find out where we should write our next message.
+        try reader.seekToEndOfNewestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+        try writer.seek(toOffset: reader.offsetInFile)
+
+        // Now that we know where to write, we need to figure out where the oldest message is.
+        try reader.seekToBeginningOfOldestMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages)
+
+        hasSetUpFileHandles = true
+    }
+
+    /// Writes message data to the cache.
+    ///
+    /// For caches that do not roll, the message data is written in the following format:
+    /// `[messageData][endOfNewestMessageMarker]`
+    /// - `messageData` is an `EncodableMessage`'s encoded data.
+    /// - `endOfNewestMessageMarker` is a big-endian encoded `MessageSpan` of length `messageSpanLength`.
+    ///
+    /// For caches that roll, the message data is written in the following format:
+    /// `[messageData][endOfNewestMessageMarker][offsetInFileOfOldestMessage]`
+    /// - `messageData` is an `EncodableMessage`'s encoded data.
+    /// - `endOfNewestMessageMarker` is a big-endian encoded `MessageSpan` of length `messageSpanLength`.
+    /// - `offsetInFileOfOldestMessage` is a big-endian encoded `Bytes` of length `oldestMessageOffsetLength`.
+    ///
+    /// By the time this method returns, the `writer`'s `offsetInFile` is always set to the
+    /// beginning of the written `endOfNewestMessageMarker`, such that the next message
+    /// written will overwrite the marker.
+    ///
+    /// - Parameter messageData: an `EncodableMessage`'s encoded data. Must be smaller than both `maximumBytes` and `MessageSpan.max`.
+    private func write(messageData: Data) throws {
```

Comment:
Right now both `append(message:)` and `write(messageData:)` have different behavior based on the value `shouldOverwriteOldMessages`. I wonder if there would be a way to refactor to avoid checking `shouldOverwriteOldMessages` in this lower-level method. IMHO checking `shouldOverwriteOldMessages` in both places makes it a bit harder to reason about since it's harder to draw to think about the code linearly (you need to keep conditionals in your head).

Reply:
Maybe it it's as simple as breaking this method apart into multiple.

Reply:
I was able to pull the common code-paths out earlier into the `write(message: T)` method and make this method more targeted.